processing/data_collection/gazette/spiders/ap_macapa.py

```diff
@@ -0,0 +1,72 @@
+import datetime as dt
+
+import scrapy
+
+from gazette.items import Gazette
+from gazette.spiders.base import BaseGazetteSpider
+
+
+class ApMacapaSpider(BaseGazetteSpider):
+
+    name = "ap_macapa"
+    allowed_domains = ["macapa.ap.gov.br"]
+    start_date = None
+
+    TERRITORY_ID = "1600303"
+
+    def __init__(self, start_date=None, end_date=None, *args, **kwargs):
+        self.start_date = dt.date(2018, 1, 1)
+        self.end_date = dt.date.today()
+
+        super(ApMacapaSpider, self).__init__(start_date, end_date)
+
+        self.logger.debug(
+            "Start date is {date}".format(date=self.start_date.isoformat())
+        )
+        self.logger.debug("End date is {date}".format(date=self.end_date.isoformat()))
+
+    def start_requests(self):
+        base_url = "https://macapa.ap.gov.br/"
+
+        target_date = self.start_date
+        data = {
+            "s": "",
+            "post_type": "official_diaries",
+            "search": "official_diaries",
+            "official_diary_number": "",
+        }
+        while target_date <= self.end_date:
+            formatted_date = target_date.strftime("%d/%m/%Y")
+            data.update(
+                {
+                    "official_diary_initial_date": formatted_date,
+                    "official_diary_final_date": formatted_date,
+                }
+            )
+
+            yield scrapy.FormRequest(
+                url=base_url, formdata=data, method="GET", meta={"date": target_date}
+            )
+            target_date = target_date + dt.timedelta(days=1)
```

Comment:
If you can define the date range, why return one request per day? Why not just one single request with the date range that you want? Image that the user wants de files from a month. In this version, you will send request 30~31 requests instead of 1.

Reply:
That is explained [here](https://github.com/okfn-brasil/querido-diario/pull/254#issuecomment-702490394), but in more detail:

That's an issue with the site: There's no reliable way to identify the correct date of the gazette by looking at the HTML code:
While most of the gazettes do have a date written on the posting, some do not:
![Screen Shot 2020-10-08 at 09 55 16](https://user-images.githubusercontent.com/10242829/95461424-98c2cd80-094c-11eb-8b5d-cf5b2f44c330.png)
And the code for these entries look like this:
```html
<div class="panel panel-default">
  <div class="panel-heading"><h4>Diário N° 3245</h4></div>
  <div class="panel-body">
    <a href="https://macapa.ap.gov.br/portal/wp-content/uploads/2019/02/DOM_3245_2018.pdf" target="_blank">
        <i class="fa fa-file-pdf-o" aria-hidden="true"></i><h4>DIÁRIO OFICIAL - N° 3245/ 2018</h4>
    </a>
  </div>
</div>
```
We can clearly see a date indication on the URL, but this is a clearly Wordpress website, and the date within the path reflects the date on which the file has been uploaded to the blog, not the date on which the gazette was issued. I did look around and found some instances where these date do not match.

If you can think of a better way to find a reliable date, please let me know.

Reply:
I understand now. Thanks for the explanation and the links.